### ISSUE TITLE
Fix S4056 FN: Localization overloads requiring more than IFormatProvider do not raise an issue

### ISSUE THREAD

**Issue by kselanne** [SonarSource/sonar-dotnet#6752]
### Description

The rule [S4056](https://sonarsource.github.io/rspec/#/rspec/S4056/csharp) fails to detect localization overloads that require further arguments beyond an `IFormatProvider`.

In particular, this concerns [`Single.TryParse`](https://learn.microsoft.com/en-us/dotnet/api/system.single.tryparse) and [`Double.TryParse`](https://learn.microsoft.com/en-us/dotnet/api/system.double.tryparse) prior to .NET 7 [1], since their localized overloads additionally require a `NumberStyles` parameter. For these methods, not specifying the culture assumes the current culture, which leads to incorrectly parsing a decimal number when it is stored invariantly with a decimal point but read in another culture using the decimal comma.

[1] .NET 7 has a localization overload that works with an `IFormatProvider` alone, so presumably a warning is correctly raised in that case.

### Repro steps

C# Code:
```csharp
using System.Globalization;

namespace FloatTryParseExample;

public static class FloatTryParseExample {
    public static void Main(string[] args) {
        var decimalPointCulture = CultureInfo.GetCultureInfo("en-US");
        var decimalCommaCulture = CultureInfo.GetCultureInfo("de-DE");

        ParseSampleFloatWithCulture("1.11", decimalPointCulture);
        ParseSampleFloatWithCulture("2.22", decimalCommaCulture);
    }

    private static void ParseSampleFloatWithCulture(string decimalPointString, CultureInfo currentCulture) {
        CultureInfo.CurrentCulture = currentCulture;

        // Non-compliant call, yet no issue raised
        float.TryParse(decimalPointString, out var decimalPointValueWithImplicitCurrentCulture);

        // Compliant call for reference
        float.TryParse(decimalPointString, NumberStyles.AllowDecimalPoint, CultureInfo.InvariantCulture, out var decimalPointValueWithExplicitInvariantCulture);

        Console.WriteLine($"Current culture: {CultureInfo.CurrentCulture}");
        Console.WriteLine($"Parse with implicit current culture: {decimalPointString} => {decimalPointValueWithImplicitCurrentCulture}");
        Console.WriteLine($"Parse with explicit invariant culture: {decimalPointString} => {decimalPointValueWithExplicitInvariantCulture}");
        Console.WriteLine();
    }
}
```

Output:
```
Current culture: en-US
Parse with implicit current culture: 1.11 => 1.11
Parse with explicit invariant culture: 1.11 => 1.11

Current culture: de-DE
Parse with implicit current culture: 2.22 => 222
Parse with explicit invariant culture: 2.22 => 2,22
```

### Expected behavior

The `float.TryParse` call in the code above labeled as non-compliant would raise an issue suggesting using the localized overload labeled as compliant in the code above.

### Actual behavior

No issue is raised for the `float.TryParse` call in the code above labeled as non-compliant.

This is due to `IsCompatibleOverload` in [CSharpOverloadHelper.cs#L80](https://github.com/SonarSource/sonar-dotnet/blob/cb0a0ed07ee56bdd12847f7b3b166038ede95b1c/analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpOverloadHelper.cs#L80) only looking for localization overloads taking exactly one further argument. Further relevant lines in the call stack:
- [CSharpOverloadHelper.cs#L25](https://github.com/SonarSource/sonar-dotnet/blob/cb0a0ed07ee56bdd12847f7b3b166038ede95b1c/analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpOverloadHelper.cs#L25)
- [SpecifyIFormatProviderOrCultureInfo.cs#L68](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.CSharp/Rules/SpecifyIFormatProviderOrCultureInfo.cs#L68)

### Known workarounds

None.

### Related information

* Visual Studio version 17.3.0
* .NET 6.0.8
* SonarAnalyzer 8.52.0.60960
* Operating System Windows 10 Build 19042

**Comment by kselanne** [comment]
Another concerned method is [`DateTime.TryParse`](https://learn.microsoft.com/en-us/dotnet/api/system.datetime.tryparse), whose localized overload prior to .NET 7 takes a DateTimeStyles parameter besides `IFormatProvider`.

**Comment by gregory-paidis-sonarsource** [comment]
Hello @kselanne , thank you for your comment.
Also, good job for going the extra-mile and pointing out why this is the case (it is indeed the `IsCompatibleOverload` method).

We discussed it internally and unfortunately we do not plan to fix this FN, because it will [potentiallly cause FPs](https://github.com/SonarSource/sonar-dotnet/issues/2139).

The possible combinations of more/less arguments, as well as their positioning in a method signature make this hard to reason about without context. 

Therefore, we prefer to miss some cases, as opposed to generating noise.

**Comment by bart-vmware** [comment]
@gregory-paidis-sonarsource If the goal is not to generate noise, please expand the blacklist for well-known cases where it doesn't make sense to pass culture info. For example `System.Convert.ToInt32/64`, `Bool.Parse` etc. Looking at their implementations, the culture is always ignored. There are many more types like that.

I already found the rule annoying because it often suggests pointless use of culture, but it got much worse after switching from .NET 6 to multi-targeting to .NET 6/8. In .NET 8, which contains generic math, several new overloads were added to primitive types that take a culture. That change was needed to comply with the new interfaces, but the culture parameter is always ignored. For example, `Guid.Parse`.

Now we need tons of places in the codebase to update from:
```c#
    var result = Guid.Parse(value);
```
to:
```c#
#if NET8_0_OR_GREATER
using System.Globalization;
#endif

#if NET8_0_OR_GREATER
    var result = Guid.Parse(value, CultureInfo.InvariantCulture);
#else
    var result = Guid.Parse(value);
#endif
```

which is a waste of time and pollution of the codebase.

Just scanning the method signatures is too simplistic. Resharper/Rider gets it right. It doesn't make sense to me why Sonar is so ignorant. In practice, culture is only potentially relevant when dates/times, floating point numbers/currencies, or `object` are involved. The overwhelming remaining cases should be suppressed.
